### PR TITLE
JavaScript Worker Select Strategy

### PR DESCRIPTION
**Pull request by nishantmonu51**
this PR adds a JavaScriptWorkerSelectStrategy which allows defining
arbitrary logic for selecting workers to run task using a JavaScript
function.

This gives users full control to implement complex worker selection
strategies based on task attributes.

### CODE REVIEW

**Comment by drcrallen** [comment]
This seems dangerous as it is essentially adding Java Script as an official language for core functionality. I assume the reason for this is "deploying javascript strategies is faster than deploying modules". This feels like it could introduce some complications, and lack of general capacity for unit testing, and lack of general documentation for how to get useful stuff from the parameters in java script, means that community use could be very frustrating.

If "faster than deploying modules" functionality really is required, then I propose moving this to an extension module, and making it very clear that the functionality is provided, but cannot be supported by the dev team and is intended to be a development aide rather than a production extension.

**Comment by drcrallen** [comment]
In general the java script stuff is good for rapid development of missing features. I'm good here if we can do 2 things:

1) Show a more complex example in the unit test, preferably using each of the 3 args at least once somewhere in the unit tests.
2) Put a little more documentation in the new docs about when the javascript one is the appropriate choice compared to writing a java one.

**Comment by nishantmonu51** [comment]
I will add a more elaborative example that can be used in real world.

**Comment by nishantmonu51** [comment]
done.

**Comment by drcrallen** [comment]
Tests fail under java 8 interestingly enough.

File: docs/content/configuration/indexing-service.md [hunk 1]
```diff
@@ -187,6 +187,29 @@ The workers with the least amount of tasks is assigned the task.
...
+
+
+|Property|Description|Default|
+|--------|-----------|-------|
+|`type`|`javascript`.|javascript|
+|`function`|String representing javascript function||
+
+Example: a function that sends batch_index_task to workers 10.0.0.1 and 10.0.0.2 and all other tasks to other available workers.
+
+```
```
Comment (drcrallen) [inline docs/content/configuration/indexing-service.md]: Can you format this a little cleaner?

Comment (nishantmonu51) [inline docs/content/configuration/indexing-service.md]: cant add newlines to json values.

**Comment by nishantmonu51** [comment]
modified the JS to be compatible with both Rhino and Nashorn

**Comment by drcrallen** [comment]
:+1:

**Comment by himanshug** [comment]
LGTM

**Pull request merged**